nle/scripts/read_tty.py

```diff
@@ -49,42 +73,129 @@ def getfile(filename):
+BRACES_COLOR = [11, 4]  # Output: Bright yellow, input: dark blue.
+
+
+# "Select Graphic Rendition" sequence.
+COLOR_REGEX = re.compile(
+    r"""
+(?P<begin>
+  (\\x1b\[)  # CSI (Control Sequence Introducer)
+  (?P<bright>[0-9])
+  ;3
+  (?P<color>[0-9])
+  m
+)
+(?P<data> .*?)
+(?P<end> \\x1b\[0m)
+""",
+    re.X,
+)
+
+# Generic control sequence
+CTRL_REGEX = re.compile(r"\\x1b\[ ([0-9];?) *.", re.X)
+
+# DEC Special Graphics
+DEC_REGEX = re.compile(
+    r"""
+(?P<begin> \\x1b\(0)
+(?P<data> .*?)
+(?P<end> \\x1b\(B)
+""",
+    re.X,
+)
+
+CSI_REGEX = re.compile(r"\\x1b\[")
+
+CSI_UNICODE = "⌘"
+
+
+def _colorsub(m):
+    """Substitute color in `COLOR_REGEX`."""
+    bright = int(m.group("bright"))
+    value = int(m.group("color"))
+    if bright:
+        value |= 8
+    return m.group("begin") + color(m.group("data"), value) + m.group("end")
 
-    def color(s, value):
-        return "\033[%d;3%dm%s\033[0m" % (bool(value & 8), value & ~8, s)
 
-    filename = sys.argv[1]
-    with getfile(filename) as f:
-        for timestamp, channel, data in ttyframes(f):
+def _ctrlsub(m):
+    return color(m.group(0), CTRL_COLOR)
+
+
+def _decsub(m):
+    """Substitute for `DEC_REGEX`."""
+    return (
+        color(m.group("begin"), DEC_COLOR)
+        + color(m.group("data"), DEC_DATA_COLOR)
+        + color(m.group("end"), DEC_COLOR)
+    )
+
+
+def main():
+    global FLAGS
+    FLAGS = parser.parse_args()
+
+    if not FLAGS.filename:
+        parser.print_help()
+        return
+
+    frames = [0, 0]
+    with getfile(FLAGS.filename) as f:
+        for timestamp, channel, data in ttyframes(f, tty2=not FLAGS.no_input):
+            frames[channel] += 1
+
+            if frames[0] < FLAGS.start:
+                continue
+
+            if frames[0] > FLAGS.end:
+                return
+
             if channel == 0:
-                data = str(data)[2:-1]  # Strip b' and '
-                channel = "<-"
+                arrow = "<-"
             elif channel == 1:
-                action, *_ = struct.unpack("<B", data)
-                data = action
-                channel = "->"
-
-            data = re.sub(
-                r"\\x1b\[([0-9];?)*.", lambda m: color(m.group(0), 8), str(data)
-            )
-            data = re.sub(
-                r"(\\x1b\(0)(.*?)(\\x1b\(B)",
-                lambda m: (
-                    color(m.group(1), 4) + color(m.group(2), 3) + color(m.group(3), 4)
-                ),
-                str(data),
-            )
-
-            print(
-                "%s %s%s%s%s"
-                % (
-                    datetime.datetime.fromtimestamp(timestamp),
-                    color(channel, 2),
-                    color("{", 11),
-                    data,
-                    color("}", 11),
+                char, *_ = struct.unpack("<B", data)
+                data = chr(char).encode("ascii", "backslashreplace")
+                arrow = "->"
+
+            data = str(data)[2:-1]  # Strip b' and '
```

Comment:
probably better to use a more 'proper' way of doing this that doesnt depend on how Python decides to represent a converted bytes to string?

Reply:
I agree! But what is the proper way of doing exactly that? It's not `data.encode("ascii", "backslashreplace")` ...

Reply:
StackOverflow seems to suggest doing what we are doing here ... https://stackoverflow.com/questions/54889234/how-to-decode-a-string-representation-of-a-bytes-object

Reply:
Looks like this is the way to do this ... https://stackoverflow.com/questions/16748083/suppress-print-without-b-prefix-for-bytes-in-python-3 and https://github.com/python/cpython/blob/e1b900247227dad49d8231f1d028872412230ab4/Objects/bytesobject.c#L1345 ...